File: Den.Dev.Grunt/Den.Dev.Grunt/Core/Modules/HaloInfinite/ConfigurationModule.cs
// <copyright file="ConfigurationModule.cs" company="Den Delimarsky">
// Developed by Den Delimarsky.
// Den Delimarsky licenses this file to you under the MIT license.
// See the LICENSE file in the project root for more information.
// The underlying API powering Den.Dev.Grunt is managed by Halo Studios and Microsoft. This wrapper is not endorsed by Halo Studios or Microsoft.
// </copyright>

using System.Threading.Tasks;
using Den.Dev.Grunt.Core.Foundation;
using Den.Dev.Grunt.Endpoints;
using Den.Dev.Grunt.Models;
using Den.Dev.Grunt.Models.ApiIngress;

namespace Den.Dev.Grunt.Core.Modules.HaloInfinite
{
    /// <summary>
    /// Module for configuration and endpoint discovery APIs.
    /// </summary>
    public class ConfigurationModule : ModuleBase
    {
        /// <summary>
        /// Initializes a new instance of the <see cref="ConfigurationModule"/> class.
        /// </summary>
        /// <param name="client">The client instance to use for API requests.</param>
        internal ConfigurationModule(ClientBase client)
            : base(client, HaloCoreEndpoints.SettingsOrigin)
        {
        }

        /// <summary>
        /// Gets the API settings container, which has the full list of available endpoints.
        /// </summary>
        /// <returns>If successful, returns an instance of Configuration that contains the full list of available endpoints. Otherwise, returns null.</returns>
        public async Task<HaloApiResultContainer<Configuration, RawResponseContainer>> GetApiSettingsContainer()
        {
            return await this.GetAsyncFullUrl<Configuration>(
                HaloCoreEndpoints.HaloInfiniteEndpointsEndpoint,
                useClearance: false,
                useSpartanToken: true);
        }
    }
}

File: src/node/src/modules/halo-infinite/configuration.module.ts
import { ModuleBase } from '../base/module-base';
import type { ClientBase } from '../../clients/base/client-base';
import type { HaloApiResult } from '../../models/common/api-result';
import { HALO_CORE_ENDPOINTS } from '../../endpoints/halo-core-endpoints';
import type { Configuration } from '../../models/halo-infinite/misc';

/**
 * Configuration module for discovering API endpoints.
 *
 * @example
 * ```typescript
 * // Get API configuration
 * const config = await client.configuration.getApiSettingsContainer();
 * console.log(config.result?.endpoints);
 * ```
 */
export class ConfigurationModule extends ModuleBase {
  constructor(client: ClientBase) {
    super(client, HALO_CORE_ENDPOINTS.SETTINGS_ORIGIN);
  }

  /**
   * Get the API settings/configuration container.
   *
   * Returns the list of all available API endpoints and their configurations.
   *
   * @returns API configuration
   */
  getApiSettingsContainer(): Promise<HaloApiResult<Configuration>> {
    return this.getFullUrl<Configuration>(HALO_CORE_ENDPOINTS.HALO_INFINITE_SETTINGS, {
      useSpartanToken: false,
    });
  }
}
